lib/src/components/NxToggle/__tests__/NxToggleRTL.test.tsx

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+import React from 'react';
+import '@testing-library/jest-dom';
+
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+
+import NxToggle, { Props } from '../NxToggle';
+
+describe('NxToggle', function() {
+  const simpleProps: Props = {
+    inputId: 'toggle-id',
+    isChecked: false,
+    onChange: () => {},
+    disabled: undefined,
+    children: 'Enables whales'
+  };
+
+  it('renders a <label> with an <input>', function() {
+    render(<NxToggle {...simpleProps}></NxToggle>);
+
+    const checkbox = screen.getByRole<HTMLInputElement>('switch', { name: 'Enables whales' });
+
+    expect(checkbox.tagName).toEqual('INPUT');
+
+    const label = checkbox.labels?.[0];
+
+    expect(label?.textContent).toEqual('Enables whales');
+  });
+
+  it('renders an input element with role switch with the correct attributes and classname', function() {
+    render(<NxToggle {...simpleProps}></NxToggle>);
+
+    const checkbox = screen.getByRole('switch');
+
+    expect(checkbox).toHaveAttribute('type', 'checkbox');
+    expect(checkbox).toHaveAttribute('id', 'toggle-id');
+    expect(checkbox).toHaveClass('nx-toggle__input');
+  });
+
+  it('adds classes specified with the className prop', function() {
+    const { container } = render(<NxToggle {...simpleProps} className="foo"></NxToggle>);
+
+    const label = container.querySelector('label');
+
+    expect(label).toHaveClass('foo');
+    expect(label).toHaveClass('nx-toggle');
+  });
+
+  it('calls its onChange prop when the input fires a change event', async function() {
+    const onChange = jest.fn();
+
+    render(<NxToggle {...simpleProps} onChange={onChange}></NxToggle>);
+
+    const checkbox = screen.getByRole('switch');
+
+    expect(onChange).not.toHaveBeenCalled();
+
+    await userEvent.click(checkbox);
+
+    expect(onChange).toHaveBeenCalled();
+  });
```

Comment:
I think we need a test for keyboard interaction as well.  

I'm not sure how I feel about that.  The previous testing paradigm relied on us basically saying "this component uses these browser built-in APIs and we just test the interaction with those APIs". We weren't attempting to test the behaviors of those APIs themselves (e.g, what user interactions cause a checkbox's `change` event to fire).  RTL seems to be opposed to that, but I worry that the new approach dramatically increases our testing burden as well as leaves us with a theoretical blindspot regarding non-standard input systems.  For instance, it seems we are now supposed to test everything about the user interacting with a component via mouse, and via keyboard, and via... what else? Things we might not even imagine.  Whereas before, we were just testing more abstractly "what happens when this component is activated" trusting the browser to send that activation signal at the appropriate time not matter what input device the user is using, even if it's one we aren't familiar with.  Thoughts @sonatype/team-plaid ?

Reply:
By mocking out the interactions between the APIs before you could easily have had something broken not realizing that a certain event isn't fired by certain user interactions right? This way requires you to be a bit more explicit about how a component is interacted with, but it also means you know the component will respond correctly under the specific conditions of the test which match up to how a user will use the component.

If you don't want to test the app they way you expect the user to interact with it you could just use [fireEvent](https://testing-library.com/docs/dom-testing-library/api-events/#fireevent) instead. `fireEvent.change` for example should trigger the `onChange` event without worrying about how the user would actually cause that event to fire. The downside to doing that though is you don't necessarily know that your event handler responds to the change event correctly for both a keyboard activation and a mouse or pointer click.

Reply:
> If you don't want to test the app they way you expect the user to interact with it you could just use [fireEvent](https://testing-library.com/docs/dom-testing-library/api-events/#fireevent) instead. fireEvent.change for example should trigger the onChange event without worrying about how the user would actually cause that event to fire. The downside to doing that though is you don't necessarily know that your event handler responds to the change event correctly for both a keyboard activation and a mouse or pointer click.

Does that actually work? I thought what @entroform was demonstrating here was that RTL doesn't work that way

Reply:
Should work just fine. We still use `fireEvent.change` in the nxrm codebase though we've been switching to userEvent over time since that more accurately simulates the browser events. For example I have this little helper function for changing form fields. You can see how we had to simulate both target and currentTarget because some handlers were using one or the other (not sure if that still applies, but it did at one point).

```
  static async changeField(fieldSelector, value) {
    fireEvent.change(fieldSelector(), {
      currentTarget: {
        name: fieldSelector().name,
        value
      },
      target: {
        name: fieldSelector().name,
        value
      }
    });
    try {
      await waitFor(() => expect(fieldSelector()).toHaveValue(value));
    } catch(error) {
      throw new Error(`${fieldSelector().name} with value ${fieldSelector().value} did not match the expected value \n ${error.message}`);
    }
  }
```

Reply:
> I think we just always used the `click` event to trigger the checkboxes. From that issue it sounds like that might be intentional on the part of the react testing library which is wierd considering the browser supports the change event on checkbox inputs. It might be a limitation of jsdom.
> 
> EDIT: I wonder if a change event sending the spacebar or firing a keypress for spacebar would trigger the checkbox?

Hmm it looks like firing keypress or change event for spacebar didn't trigger the checkbox.
I also tried it using `userEvent.keyboard` and didn't work. I read that it might be JSDOM limitation as well

```
await checkbox.focus();
userEvent.keyboard('{Space}');
```